### ISSUE TITLE
feat: Auto-tuning kernel dispatcher with size-aware selection

### ISSUE THREAD

**Issue by m96-chan** [m96-chan/PyGPUkit#120]
## Summary

Implement an auto-tuning dispatcher that benchmarks kernels at startup and automatically selects the optimal implementation based on GPU model AND operation dimensions.

## Motivation

- Different GPUs (SM80/86/89/90/120) have different optimal kernels
- LLM inference is **GEMV-heavy** during decode (M=1), not GEMM
- Optimal kernel varies by matrix dimensions
- Manual kernel selection is error-prone and suboptimal

## Key Insight: Size-Aware Dispatch

The optimal kernel changes based on matrix dimensions:

| M (batch/seq) | Workload | Optimal Strategy |
|---------------|----------|------------------|
| M=1 | GEMV | Memory bandwidth optimized |
| M=2-8 | Small GEMM | Batched GEMV or thin GEMM |
| M=16-128 | Medium GEMM | TensorCore with small tiles |
| M=256+ | Large GEMM | Full TensorCore GEMM |

## Proposed API

```python
import pygpukit

# Option 1: Auto-tune at init (recommended)
pygpukit.init(autotune=True)  # ~10s benchmark on first run

# Option 2: Manual tuning
pygpukit.autotune(
    operations=["gemm", "gemv", "attention"],
    sizes=[(1, 4096, 4096), (8, 4096, 4096), ...],
    cache_path="~/.pygpukit/bench_cache.json"
)

# Transparent dispatch - best kernel auto-selected
C = A @ B  # Dispatcher picks optimal kernel based on shapes
```

## Dispatcher Architecture

```
          ┌─────────────────────────────────────┐
          │         Kernel Dispatcher           │
          └─────────────────────────────────────┘
                          │
          ┌───────────────┼───────────────┐
          ▼               ▼               ▼
    ┌──────────┐   ┌──────────┐   ┌──────────┐
    │ Size     │   │ GPU      │   │ Dtype    │
    │ Buckets  │   │ SM arch  │   │ Match    │
    └──────────┘   └──────────┘   └──────────┘
          │               │               │
          └───────────────┼───────────────┘
                          ▼
          ┌─────────────────────────────────────┐
          │     Lookup Table (from cache)       │
          │  (sm, dtype, op, size_bucket) → fn  │
          └─────────────────────────────────────┘
                          │
                          ▼
          ┌─────────────────────────────────────┐
          │         Execute Best Kernel         │
          └─────────────────────────────────────┘
```

### GEMV (Decode Phase - Most Critical!)
| Implementation | Notes |
|----------------|-------|
| cuBLAS SGEMV/HGEMV | Baseline |
| cuBLASLt M=1 GEMM | Descriptor caching |
| Custom GEMV | Bandwidth-optimized |
| FP8 GEMV | SM90/120 native |
| Batched GEMV | Multi-layer fusion |
| Persistent GEMV | Reduce launch overhead |

### GEMM (Prefill Phase)
| Implementation | Notes |
|----------------|-------|
| cuBLASLt | General purpose |
| CUTLASS | Template-based |
| Custom TF32 | Ampere optimized |
| FP8 GEMM | SM90/120 |

### Attention
| Implementation | Notes |
|----------------|-------|
| FlashAttention v2 | Memory efficient |
| SDPA (PyTorch-style) | Fallback |
| PagedAttention | vLLM-style |
| Custom fused | RoPE + Attention |

### Other Ops
- RMSNorm: fused vs separate
- RoPE: fused vs separate
- SwiGLU: fused gate*up vs separate
- Softmax: online vs two-pass

## Size Buckets

```python
GEMV_BUCKETS = [1]  # M=1 only
SMALL_GEMM_BUCKETS = [2, 4, 8, 16]
MEDIUM_GEMM_BUCKETS = [32, 64, 128, 256]
LARGE_GEMM_BUCKETS = [512, 1024, 2048, 4096]
```

## Cache Format

```json
{
  "version": 1,
  "gpu": "NVIDIA GeForce RTX 5090",
  "sm": 120,
  "driver": "572.xx",
  "timestamp": "2025-12-26T12:00:00Z",
  "benchmarks": {
    "gemv_bf16": {
      "1x4096x4096": {"winner": "custom_gemv", "bandwidth_gbps": 1800},
      "1x4096x14336": {"winner": "cublaslt", "bandwidth_gbps": 1750}
    },
    "gemm_bf16": {
      "8x4096x4096": {"winner": "cublaslt", "tflops": 45},
      "128x4096x4096": {"winner": "cutlass", "tflops": 280},
      "4096x4096x4096": {"winner": "cutlass", "tflops": 316}
    },
    "gemm_fp8": {
      "128x4096x4096": {"winner": "cutlass_sm120", "tflops": 580}
    },
    "attention_causal": {
      "1x32x8192": {"winner": "flash_v2", "ms": 0.8},
      "128x32x8192": {"winner": "flash_v2", "ms": 12.4}
    }
  }
}
```

### Phase 1: Infrastructure
- [ ] Kernel registry with metadata
- [ ] Size bucket classification
- [ ] Cache file format and I/O
- [ ] Benchmark harness

### Phase 2: GEMV/GEMM Dispatch
- [ ] GEMV implementations and benchmarks
- [ ] GEMM implementations and benchmarks
- [ ] Size-aware dispatch logic
- [ ] cuBLASLt descriptor pool

### Phase 3: Full Operator Coverage
- [ ] Attention variants
- [ ] Normalization (LayerNorm, RMSNorm)
- [ ] Activation functions (SwiGLU, GELU)
- [ ] RoPE embedding

### Phase 4: Advanced
- [ ] Runtime re-tuning (detect perf regression)
- [ ] Multi-GPU dispatch
- [ ] Fusion opportunity detection
- [ ] Profile-guided optimization

## Target Version

v0.3.0

## Related Issues

- #118 FP8 model loading
- #119 Auto-quantization
